benchmarks/src/main/kotlin/bfschannel/BfsChannelBenchmark.kt

```diff
@@ -0,0 +1,363 @@
+    val end = System.nanoTime()
+
+    return ExecutionResults(end - start, distances)
+}
+
+private fun runSequentialBfs(startNode: Node, nodes: Int) : ExecutionResults {
+    val start = System.nanoTime()
+    val distances = bfsSequential(startNode, nodes)
+    val end = System.nanoTime()
+
+    return ExecutionResults(end - start, distances)
+}
+
+class ExecutionResults(val executionTime : Long, val distances : List<Long>)
+
+fun bfsSequential(startNode: Node, nodes: Int): List<Long> {
+    val distances = Array(nodes) { Long.MAX_VALUE }
+    // The distance to the start node is `0`
+    distances[startNode.id] = 0
+
+    val queue = LinkedList<Node>()
+    queue.add(startNode)
+    while (queue.isNotEmpty()) {
+        val currentNode = queue.poll()
+        for (edge in currentNode.outgoingEdges) {
+            val newDistance = distances[currentNode.id] + 1
+            val oldDistance = distances[edge.to.id]
+            if (newDistance < oldDistance) {
+                distances[edge.to.id] = newDistance
+                queue.add(edge.to)
+            }
+        }
+    }
+    return distances.toList()
+}
+
+fun bfsParallel(start: Node, nodes: Int, coroutines: Int): List<Long> {
+    val distances = Array(nodes) { AtomicLong(Long.MAX_VALUE) }
+    // The distance to the start node is `0`
+    distances[start.id].set(0)
+
+    val queue : Channel<Node> = SelfClosingChannel(coroutines)
+    queue.offer(start)
+    // Run worker threads and wait until the total work is done
+    val jobs = ArrayList<Job>()
+    repeat(coroutines) {
+        jobs += GlobalScope.launch {
+            while (true) {
+                val currentNode = queue.receiveOrClosed().valueOrNull ?: break
+                for (edge in currentNode.outgoingEdges) {
+                    relaxEdge(distances, currentNode, edge.to, queue)
+                }
+            }
+        }
+    }
+
+    runBlocking {
+        for (job in jobs) {
+            job.join()
+        }
+    }
+    // Return the result
+    return distances.map { long -> long.toLong() }
+}
+
+private fun relaxEdge(distances: Array<AtomicLong>, currentNode: Node, nextNode: Node, queue: Channel<Node>) {
+    val newDistance = distances[currentNode.id].get() + 1
+    // try to compare and set if new distance is less than the old one
+    while (true) {
+        val toDistance = distances[nextNode.id].get()
+        if (toDistance <= newDistance) {
+            return
+        }
+        if (distances[nextNode.id].compareAndSet(toDistance, newDistance)) {
+            queue.offer(nextNode)
+            return
+        }
+    }
+}
+
+class Node(val id: Int) {
+    private val _outgoingEdges = arrayListOf<Edge>()
+    val outgoingEdges: List<Edge> = _outgoingEdges
+
+    fun addEdge(edge: Edge) {
+        _outgoingEdges.add(edge)
+    }
+}
+
+data class Edge(val to: Node)
+
+/**
+ * Channel that closes itself if specified amount of coroutines start waiting on receive
```

Comment:
"This channel implementation does not suspend on sends and closes itself if the number of waiting receivers exceeds [maxWaitingReceivers]."